swift/apple/FirezoneNetworkExtension/Adapter.swift

```diff
@@ -389,56 +340,26 @@ extension Adapter: CallbackHandlerDelegate {
     workQueue.async { [weak self] in
       guard let self = self else { return }
 
-      self.logger.log("Adapter.onSetInterfaceConfig")
-
-      switch self.state {
-      case .startingTunnel:
-        self.networkSettings = NetworkSettings(
-          tunnelAddressIPv4: tunnelAddressIPv4, tunnelAddressIPv6: tunnelAddressIPv6,
-          dnsAddresses: dnsAddresses)
-      case .tunnelReady:
-        if let networkSettings = self.networkSettings {
-          networkSettings.apply(
-            on: packetTunnelProvider,
-            logger: self.logger,
-            completionHandler: nil
-          )
-        }
+      logger.log(
+        "Adapter.onSetInterfaceConfig: \(tunnelAddressIPv4) \(tunnelAddressIPv6) \(dnsAddresses)")
 
-      case .stoppingTunnel, .stoppedTunnel, .stoppingTunnelTemporarily, .stoppedTunnelTemporarily:
-        // This is not expected to happen
-        break
-      }
-    }
-  }
-
-  public func onTunnelReady() {
-    workQueue.async { [weak self] in
-      guard let self = self else { return }
-
-      self.logger.log("Adapter.onTunnelReady")
-      guard case .startingTunnel(let session, let onStarted) = self.state else {
-        self.logger.error(
-          "Adapter.onTunnelReady: Unexpected state: \(self.state)")
-        return
-      }
-      guard let networkSettings = self.networkSettings else {
-        self.logger.error("Adapter.onTunnelReady: No network settings")
-        return
-      }
-
-      // Connlib's up, set it as the default DNS
-      networkSettings.setMatchDomains([""])
-      networkSettings.apply(on: packetTunnelProvider, logger: self.logger) { error in
-        if let error = error {
-          self.logger.error("\(#function): \(error)")
-          onStarted?(AdapterError.setNetworkSettings(error))
-          self.state = .stoppedTunnel
-        } else {
-          onStarted?(nil)
-          self.state = .tunnelReady(session: session)
-          self.beginPathMonitoring()
-        }
+      switch state {
+      case .startingTunnel(let session, let onStarted):
+        onStarted?(nil)
+        networkSettings.tunnelAddressIPv4 = tunnelAddressIPv4
+        networkSettings.tunnelAddressIPv6 = tunnelAddressIPv6
+        networkSettings.dnsAddresses = dnsAddresses
+        networkSettings.apply()
+        state = .tunnelReady(session: session)
+        beginPathMonitoring()
+      case .tunnelReady(session: _):
+        networkSettings.tunnelAddressIPv4 = tunnelAddressIPv4
+        networkSettings.tunnelAddressIPv6 = tunnelAddressIPv6
+        networkSettings.dnsAddresses = dnsAddresses
+        networkSettings.apply()
+      case .stoppedTunnel:
+        logger.error(
+          "\(#function): Unexpected state: \(self.state)")
```

Comment:
We could probably even collapse these states further into just `.startedTunnel` and `.stoppedTunnel`. There's not much that can go wrong between `.startingTunnel` and `.tunnelReady`.

Reply:
Do you have a timeout for going into `tunnelReady`? If not then the state is already pointless I'd say.

All of these states should be able to be removed. Either you have a pointer or you don't. `connect` gives you the pointer, `disconnect` "consumes" it. `on_disconnected` you should throw it away.

Your state machine should just be an `Option` or whatever the equivalent of that is in Swift :)

Reply:
`on_disconnected` is used to notify the client their session has ended (portal 401), but otherwise I'm with you.

We have to call a completionHandler when the tunnel "is up" but that could just be right when first `onSetInterfaceConfig` happens, like we do now.

Reply:
> `on_disconnected` is used to notify the client their session has ended (portal 401)

It is also used when connlib crashes. I think the main semantics here is, once you receive `on_disconnected`, the pointer you had to connlib is invalid and should no longer be used. Why it was disconnected is in the error which at the moment is unfortunately untyped.

Reply:
> We have to call a completionHandler when the tunnel "is up" but that could just be right when first `onSetInterfaceConfig` happens, like we do now.

Who mandates this completion handler? Could you call it right after you initialized connlib using `connect`?

Reply:
> Who mandates this completion handler? Could you call it right after you initialized connlib using connect?

We could, but we need interface IPs at a minimum to bring the interface up, so that's why I wait for the `onSetInterfaceConfig`.

Once the completionHandler is called, Apple considers the interface active and starts sending packets to it.

Reply:
> Do you have a timeout for going into tunnelReady? If not then the state is already pointless I'd say.

Not on the Apple side -- I think it just relies on connlib's callback. Yeah let me take another stab at cleaning this up some more! :D

Reply:
Done, removed the pointless `.startingTunnel` state.

Reply:
https://github.com/firezone/firezone/pull/4133/commits/bd4d248c547ba15eea071189317c33cf939db9b5